gapis/api/vulkan/frame_loop.go

```diff
@@ -1360,6 +1414,273 @@ func (f *frameLoop) resetResources(ctx context.Context, stateBuilder *stateBuild
 	return nil
 }
 
+func (f *frameLoop) resetInstances(ctx context.Context, stateBuilder *stateBuilder) error {
+
+	// For every Instance that we need to destroy at the end of the loop...
+	for toDestroy := range f.instanceToDestroy {
+		// Write the command to delete the created object
+		instance := GetState(f.loopEndState).instances.Get(toDestroy)
+		stateBuilder.write(stateBuilder.cb.VkDestroyInstance(instance.VulkanHandle(), memory.Nullptr))
+	}
+
+	// For every Instance that we need to create at the end of the loop...
+	for toCreate := range f.instanceToCreate {
+		// Write the commands needed to recreate the destroyed object
+		instance := GetState(f.loopStartState).instances.Get(toCreate)
+		stateBuilder.createInstance(instance.VulkanHandle(), instance)
+	}
+
+	for _, deviceObject := range GetState(f.loopStartState).Devices().All() {
+
+		physicalDevice := deviceObject.PhysicalDevice()
+		physicalDeviceObject := GetState(f.loopStartState).PhysicalDevices().All()[physicalDevice]
+		instance := physicalDeviceObject.Instance()
+
+		if _, ok := f.instanceToCreate[instance]; ok {
+			f.deviceToDestroy[deviceObject.VulkanHandle()] = true
+			f.deviceToCreate[deviceObject.VulkanHandle()] = true
+		}
+	}
+
+	return nil
+}
+
+func (f *frameLoop) resetDevices(ctx context.Context, stateBuilder *stateBuilder) error {
+
+	// For every Device that we need to destroy at the end of the loop...
+	for toDestroy := range f.deviceToDestroy {
+		// Write the command to delete the created object
+		device := GetState(f.loopEndState).devices.Get(toDestroy)
+		stateBuilder.write(stateBuilder.cb.VkDestroyDevice(device.VulkanHandle(), memory.Nullptr))
+	}
+
+	// For every Device that we need to create at the end of the loop...
+	for toCreate := range f.deviceToCreate {
+		// Write the commands needed to recreate the destroyed object
+		device := GetState(f.loopStartState).devices.Get(toCreate)
+		stateBuilder.createDevice(device)
+	}
+
+	for _, memoryObject := range GetState(f.loopStartState).DeviceMemories().All() {
+
+		device := memoryObject.Device()
+
+		if _, ok := f.deviceToCreate[device]; ok {
+			f.memoryToFree[memoryObject.VulkanHandle()] = true
+			f.memoryToAllocate[memoryObject.VulkanHandle()] = true
+			// TODO: @renfengliu What do we need to do here with memoryToUnmap/memoryToMap/mappedAddress?
+		}
+	}
+
+	for _, bufferObject := range GetState(f.loopStartState).Buffers().All() {
```

Comment:
How about bufferview?

Reply:
BufferView and ImageView should be taken care of inductively by rebuilding the buffers and images they are built from.